src/category_theory/subobject.lean

```diff
@@ -0,0 +1,963 @@
+
+We define `subobject X` as the quotient (by isomorphisms) of
+`mono_over X := {f : over X // mono f.hom}`.
+
+Here `mono_over X` is a thin category (a pair of objects has at most one morphism between them),
+so we can think of it as a preorder. However as it is not skeletal, it is not a partial order.
+
+We provide
+* `def pullback [has_pullbacks C] (f : X ⟶ Y) : subobject Y ⥤ subobject X`
+* `def map (f : X ⟶ Y) [mono f] : subobject X ⥤ subobject Y`
+* `def «exists» [has_images C] (f : X ⟶ Y) : subobject X ⥤ subobject Y`
+(each first at the level of `mono_over`), and prove their basic properties and relationships.
+
+We also provide the `semilattice_inf_top (subobject X)` instance when `[has_pullback C]`,
+and the `semilattice_inf (subobject X)` instance when `[has_images C] [has_binary_coproducts C]`.
+
+## Notes
+
+This development originally appeared in Bhavik Mehta's "Topos theory for Lean" repository,
+and was ported to mathlib by Scott Morrison.
+
+### Implementation note
+
+Currently we describe `pullback`, `map`, etc., as functors.
+It may be better to just say that they are monotone functions,
+and even avoid using categorical language entirely when describing `subobject X`.
+(It's worth keeping this in mind in future use; it should be a relatively easy change here
+if it looks preferable.)
+
+### Relation to pseudoelements
+
+There is a separate development of pseudoelements in `category_theory.abelian.pseudoelements`,
+as a quotient (but not by isomorphism) of `over X`.
+
+When a morphism `f` has an image, it represents the same pseudoelement.
+In a category with images `pseudoelements X` could be constructed as a quotient of `mono_over X`.
+In fact, in an abelian category (I'm not sure in what generality beyond that),
+`pseudoelements X` agrees with `subobject X`, but we haven't developed this in mathlib yet.
+
+-/
+
+universes v₁ v₂ u₁ u₂
+
+noncomputable theory
+namespace category_theory
+
+open category_theory category_theory.category category_theory.limits
+
+variables {C : Type u₁} [category.{v₁} C] {X Y Z : C}
+variables {D : Type u₂} [category.{v₂} D]
+
+/--
+The category of monomorphisms into `X` as a full subcategory of the over category.
+This isn't skeletal, so it's not a partial order.
+
+Later we define `subobject X` as the quotient of this by isomorphisms.
+-/
+@[derive [category, λ t, has_coe t (over X)]]
+def mono_over (X : C) := {f : over X // mono f.hom}
+
+attribute [priority 100] mono_over.has_coe
+-- FIXME
+-- The linter still claims that `mono_over.has_coe` is a dangerous instance:
+/-
+```
+/- The `instance_priority` linter reports: -/
+/- DANGEROUS INSTANCE PRIORITIES.
+The following instances always apply, and therefore should have a priority < 1000.
+If you don't know what priority to choose, use priority 100.
+See note [lower instance priority] for instructions to change the priority.: -/
+#print category_theory.mono_over.has_coe /- LINTER FAILED:
+unknown declaration '[anonymous]' -/
+```
+-/
```

Comment:
@fpvandoorn, @robertylewis, could one of you assist here? I'm reducing the priority of this `has_coe` instance per the linter's request, but it is still upset.

I'm not sure if the `LINTER FAILED: unknown declaration '[anonymous]'` is a sign of trouble.

Reply:
The type of `mono_over.has_coe` is 
```lean
@[instance]
protected def category_theory.mono_over.has_coe : Π {C : Type u₁} [_inst_1 : category C] (X : C), 
  (λ (t : Type (max u₁ v₁)), has_coe t (over X)) (mono_over X)
```
Probably the problem is that the linter doesn't expect the type of an instance to have a beta-redex.

I'll try to fix the linter, but I think the derive handler should also be fixed to generate a type in beta-whnf.

Reply:
Though I guess `derive [λ t, has_coe t (over X)]` is asking for a beta redex.

Reply:
Thanks for the patch!

Reply:
> Though I guess `derive [λ t, has_coe t (over X)]` is asking for a beta redex.

Yeah to be honest I didn't expect this `derive` to work